File: queues/moving_average.py
from collections import deque
from typing import Deque


class MovingAverage:
    def __init__(self, size: int):
        self.size = size
        self.queue: Deque[int] = deque()
        self.sum = 0

    def next(self, val: int) -> float:
        self.queue.append(val)
        self.sum += val
        if len(self.queue) > self.size:
            self.sum -= self.queue.popleft()
        return self.sum / len(self.queue)

# Time Complexity: O(1) for each next call, as each element is added and removed at most once.
# Space Complexity: O(n) in the worst case, where n is the size of the moving average window.

File: queues/MovingAverage.java
import java.util.ArrayDeque;
import java.util.Deque;

public class MovingAverage {
  int size;
  Deque<Integer> queue;
  int sum;

  public MovingAverage(int size) {
    this.size = size;
    this.queue = new ArrayDeque<>();
    this.sum = 0;
  }

  public double next(int val) {
    sum += val;
    queue.addLast(val);
    if (queue.size() > size) {
      sum -= queue.removeFirst();
    }
    return (double) sum / queue.size();
  }
}

// Time complexity O(1) for each call to next()
// Space complexity O(size) where size is the maximum number of elements in the moving average window